File: src/stdlib/ram.ts
/**
 * `ram`: arithmetic.
 *
 * A ram is the one that does the heavy butting, so it got the maths. Baa has a
 * single numeric type (IEEE-754 double), and this module is where the integer
 * flavoured operations live: `idiv`, `gcd`, `clamp`, rounding.
 */

import { BaaError } from "../diagnostics/diagnostic.ts";
import { BaaArray } from "../runtime/values.ts";
import type { Value } from "../runtime/values.ts";
import { argArray, argNumber, defineModule, fn } from "./define.ts";

function numbersOf(fnName: string, array: BaaArray, span: import("../diagnostics/source.ts").Span): number[] {
  return array.items.map((item, index) => {
    if (typeof item !== "number") {
      throw BaaError.of("BAA311", [fnName, "an array of numbers", "1", `a ${typeof item} at index ${index}`], {
        span,
        note: "every item must be a number",
      });
    }
    return item;
  });
}

export function createRam() {
  return defineModule("ram", {
    PI: Math.PI,
    E: Math.E,
    TAU: Math.PI * 2,
    INF: Number.POSITIVE_INFINITY,
    NAN: Number.NaN,
    EPSILON: Number.EPSILON,
    MAX_SAFE_WHOLE: Number.MAX_SAFE_INTEGER,

    abs: fn(1, 1, "Absolute value.", (args, ctx) => Math.abs(argNumber("ram.abs", args, 0, ctx.span))),
    sign: fn(1, 1, "-1, 0 or 1.", (args, ctx) => Math.sign(argNumber("ram.sign", args, 0, ctx.span))),
    floor: fn(1, 1, "Round down.", (args, ctx) => Math.floor(argNumber("ram.floor", args, 0, ctx.span))),
    ceil: fn(1, 1, "Round up.", (args, ctx) => Math.ceil(argNumber("ram.ceil", args, 0, ctx.span))),
    trunc: fn(1, 1, "Drop the fractional part.", (args, ctx) =>
      Math.trunc(argNumber("ram.trunc", args, 0, ctx.span)),
    ),
    round: fn(1, 2, "Round to the nearest whole number, or to n decimal places.", (args, ctx) => {
      const value = argNumber("ram.round", args, 0, ctx.span);
      if (args.length === 1) return Math.round(value);
      const digits = argNumber("ram.round", args, 1, ctx.span);
      const factor = 10 ** digits;
      return Math.round(value * factor) / factor;
    }),

    sqrt: fn(1, 1, "Square root.", (args, ctx) => {
      const value = argNumber("ram.sqrt", args, 0, ctx.span);
      if (value < 0) {
        throw BaaError.of("BAA301", ["ram.sqrt of a negative number is not a real number"], {
          span: ctx.span,
          note: "negative input",
          help: "Check the sign first, or use `ram.abs(x)`.",
        });
      }
      return Math.sqrt(value);
    }),
    pow: fn(2, 2, "Raise to a power.", (args, ctx) =>
      argNumber("ram.pow", args, 0, ctx.span) ** argNumber("ram.pow", args, 1, ctx.span),
    ),
    exp: fn(1, 1, "e raised to a power.", (args, ctx) => Math.exp(argNumber("ram.exp", args, 0, ctx.span))),
    log: fn(1, 2, "Natural logarithm, or logarithm in a given base.", (args, ctx) => {
      const value = argNumber("ram.log", args, 0, ctx.span);
      if (value <= 0) {
        throw BaaError.of("BAA301", ["ram.log needs a value greater than zero"], {
          span: ctx.span,
          note: "out of domain",
        });
      }
      if (args.length === 1) return Math.log(value);
      return Math.log(value) / Math.log(argNumber("ram.log", args, 1, ctx.span));
    }),

    sin: fn(1, 1, "Sine, in radians.", (args, ctx) => Math.sin(argNumber("ram.sin", args, 0, ctx.span))),
    cos: fn(1, 1, "Cosine, in radians.", (args, ctx) => Math.cos(argNumber("ram.cos", args, 0, ctx.span))),
    tan: fn(1, 1, "Tangent, in radians.", (args, ctx) => Math.tan(argNumber("ram.tan", args, 0, ctx.span))),
    atan2: fn(2, 2, "Angle of the vector (x, y), in radians.", (args, ctx) =>
      Math.atan2(argNumber("ram.atan2", args, 0, ctx.span), argNumber("ram.atan2", args, 1, ctx.span)),
    ),
    hypot: fn(2, 2, "Length of the vector (x, y).", (args, ctx) =>
      Math.hypot(argNumber("ram.hypot", args, 0, ctx.span), argNumber("ram.hypot", args, 1, ctx.span)),
    ),

    min: fn(1, Number.MAX_SAFE_INTEGER, "Smallest of the arguments.", (args, ctx) =>
      Math.min(...args.map((_, index) => argNumber("ram.min", args, index, ctx.span))),
    ),
    max: fn(1, Number.MAX_SAFE_INTEGER, "Largest of the arguments.", (args, ctx) =>
      Math.max(...args.map((_, index) => argNumber("ram.max", args, index, ctx.span))),
    ),
    clamp: fn(3, 3, "Constrain a value between low and high.", (args, ctx) => {
      const value = argNumber("ram.clamp", args, 0, ctx.span);
      const low = argNumber("ram.clamp", args, 1, ctx.span);
      const high = argNumber("ram.clamp", args, 2, ctx.span);
      if (low > high) {
        throw BaaError.of("BAA311", ["ram.clamp", "low <= high", "2", "a low bound above the high bound"], {
          span: ctx.span,
        });
      }
      return Math.min(Math.max(value, low), high);
    }),
    lerp: fn(3, 3, "Linear interpolation between a and b.", (args, ctx) => {
      const a = argNumber("ram.lerp", args, 0, ctx.span);
      const b = argNumber("ram.lerp", args, 1, ctx.span);
      const t = argNumber("ram.lerp", args, 2, ctx.span);
      return a + (b - a) * t;
    }),

    idiv: fn(2, 2, "Integer division, rounding toward negative infinity.", (args, ctx) => {
      const a = argNumber("ram.idiv", args, 0, ctx.span);
      const b = argNumber("ram.idiv", args, 1, ctx.span);
      if (b === 0) throw BaaError.of("BAA306", [], { span: ctx.span, note: "divisor is zero" });
      return Math.floor(a / b);
    }),
    modulo: fn(2, 2, "Remainder that always takes the sign of the divisor.", (args, ctx) => {
      const a = argNumber("ram.modulo", args, 0, ctx.span);
      const b = argNumber("ram.modulo", args, 1, ctx.span);
      if (b === 0) throw BaaError.of("BAA306", [], { span: ctx.span, note: "divisor is zero" });
      return ((a % b) + b) % b;
    }),
    gcd: fn(2, 2, "Greatest common divisor of two whole numbers.", (args, ctx) => {
      let a = Math.abs(argNumber("ram.gcd", args, 0, ctx.span));
      let b = Math.abs(argNumber("ram.gcd", args, 1, ctx.span));
      while (b > 0) {
        const next = a % b;
        a = b;
        b = next;
      }
      return a;
    }),

    is_nan: fn(1, 1, "True when the value is not a number.", (args) =>
      typeof args[0] === "number" && Number.isNaN(args[0]),
    ),
    is_finite: fn(1, 1, "True when the value is a finite number.", (args) =>
      typeof args[0] === "number" && Number.isFinite(args[0]),
    ),
    is_whole: fn(1, 1, "True when the value is a whole number.", (args) =>
      typeof args[0] === "number" && Number.isInteger(args[0]),
    ),

    sum: fn(1, 1, "Add every number in an array.", (args, ctx) =>
      numbersOf("ram.sum", argArray("ram.sum", args, 0, ctx.span), ctx.span).reduce(
        (total, value) => total + value,
        0,
      ),
    ),
    mean: fn(1, 1, "Arithmetic mean of an array, or nil when empty.", (args, ctx) => {
      const numbers = numbersOf("ram.mean", argArray("ram.mean", args, 0, ctx.span), ctx.span);
      if (numbers.length === 0) return null;
      return numbers.reduce((total, value) => total + value, 0) / numbers.length;
    }),
    median: fn(1, 1, "Median of an array, or nil when empty.", (args, ctx) => {
      const numbers = numbersOf("ram.median", argArray("ram.median", args, 0, ctx.span), ctx.span)
        .slice()
        .sort((a, b) => a - b);
      if (numbers.length === 0) return null;
      const middle = Math.floor(numbers.length / 2);
      return numbers.length % 2 === 1
        ? numbers[middle]!
        : (numbers[middle - 1]! + numbers[middle]!) / 2;
    }),

    to_binary: fn(1, 1, "Binary text for a whole number.", (args, ctx) =>
      wholeToString("ram.to_binary", args, ctx.span, 2),
    ),
    to_hex: fn(1, 1, "Hexadecimal text for a whole number.", (args, ctx) =>
      wholeToString("ram.to_hex", args, ctx.span, 16),
    ),
    parse: fn(1, 2, "Parse text as a number in a given base (default 10).", (args, ctx) => {
      const text = args[0];
      if (typeof text !== "string") {
        throw BaaError.of("BAA311", ["ram.parse", "a string", "1", typeof text], { span: ctx.span });
      }
      const base = args.length > 1 ? argNumber("ram.parse", args, 1, ctx.span) : 10;
      const trimmed = text.trim();
      // `Number("")` is 0, which would make blank text parse as a number.
      if (trimmed === "") return null;
      const value = base === 10 ? Number(trimmed) : Number.parseInt(trimmed, base);
      return Number.isNaN(value) ? null : value;
    }),
  });
}

function wholeToString(
  fnName: string,
  args: Value[],
  span: import("../diagnostics/source.ts").Span,
  radix: number,
): string {
  const value = argNumber(fnName, args, 0, span);
  if (!Number.isInteger(value)) {
    throw BaaError.of("BAA311", [fnName, "a whole number", "1", "a fraction"], { span });
  }
  return value.toString(radix);
}

File: rust/crates/baa-native/src/stdlib/ram.rs
//! `ram`: arithmetic. A port of `src/stdlib/ram.ts`.
//!
//! Every function here exists there with the same name, arity and edge-case
//! behaviour, including the ones that are easy to get subtly different:
//! `modulo` takes the sign of its divisor while `%` takes the sign of its
//! left operand, `idiv` rounds towards negative infinity rather than zero, and
//! `sqrt` of a negative number is an error rather than a quiet `nan`.

use std::rc::Rc;

use crate::ast::Span;
use crate::interp::{Flow, Interpreter, Res};
use crate::number;
use crate::value::{Module, Native, Value};

use super::{module_of, need_array, need_number, need_string};

const FUNCTIONS: &[(&str, usize, usize)] = &[
    ("abs", 1, 1),
    ("sign", 1, 1),
    ("floor", 1, 1),
    ("ceil", 1, 1),
    ("trunc", 1, 1),
    ("round", 1, 2),
    ("sqrt", 1, 1),
    ("pow", 2, 2),
    ("exp", 1, 1),
    ("log", 1, 2),
    ("sin", 1, 1),
    ("cos", 1, 1),
    ("tan", 1, 1),
    ("atan2", 2, 2),
    ("hypot", 2, 2),
    ("min", 1, usize::MAX),
    ("max", 1, usize::MAX),
    ("clamp", 3, 3),
    ("lerp", 3, 3),
    ("idiv", 2, 2),
    ("modulo", 2, 2),
    ("gcd", 2, 2),
    ("is_nan", 1, 1),
    ("is_finite", 1, 1),
    ("is_whole", 1, 1),
    ("sum", 1, 1),
    ("mean", 1, 1),
    ("median", 1, 1),
    ("to_binary", 1, 1),
    ("to_hex", 1, 1),
    ("parse", 1, 2),
];

pub fn module() -> Rc<Module> {
    let mut exports: Vec<(&str, Value)> = vec![
        ("PI", Value::Number(std::f64::consts::PI)),
        ("E", Value::Number(std::f64::consts::E)),
        ("TAU", Value::Number(std::f64::consts::PI * 2.0)),
        ("INF", Value::Number(f64::INFINITY)),
        ("NAN", Value::Number(f64::NAN)),
        ("EPSILON", Value::Number(f64::EPSILON)),
        ("MAX_SAFE_WHOLE", Value::Number(9_007_199_254_740_991.0)),
    ];
    for (name, min, max) in FUNCTIONS {
        exports.push((
            name,
            Value::Native(Rc::new(Native {
                name: format!("ram.{name}"),
                method: name,
                min_args: *min,
                max_args: *max,
                call,
                receiver: None,
            })),
        ));
    }
    module_of("ram", exports)
}

fn call(interp: &mut Interpreter, native: &Native, args: Vec<Value>, span: Span) -> Res<Value> {
    let name = &native.name;
    let one = |interp: &Interpreter| need_number(interp, name, &args, 0, span);

    Ok(match native.method {
        "abs" => Value::Number(one(interp)?.abs()),
        // `Math.sign` returns the value itself for NaN and preserves -0, which
        // `f64::signum` does not: signum(-0.0) is -1.0 and signum(NaN) is NaN
        // with a sign. Written out rather than borrowed.
        "sign" => {
            let value = one(interp)?;
            Value::Number(if value.is_nan() {
                f64::NAN
            } else if value > 0.0 {
                1.0
            } else if value < 0.0 {
                -1.0
            } else {
                value
            })
        }
        "floor" => Value::Number(one(interp)?.floor()),
        "ceil" => Value::Number(one(interp)?.ceil()),
        "trunc" => Value::Number(one(interp)?.trunc()),
        "round" => {
            let value = one(interp)?;
            if args.len() == 1 {
                // JavaScript rounds halves towards positive infinity.
                Value::Number((value + 0.5).floor())
            } else {
                let digits = need_number(interp, name, &args, 1, span)?;
                let factor = 10f64.powf(digits);
                Value::Number((value * factor + 0.5).floor() / factor)
            }
        }
        "sqrt" => {
            let value = one(interp)?;
            if value < 0.0 {
                return Err(Flow::Err(
                    interp
                        .error(
                            "BAA301",
                            vec!["ram.sqrt of a negative number is not a real number".into()],
                            span,
                        )
                        .with_note("negative input")
                        .with_help("Check the sign first, or use `ram.abs(x)`."),
                ));
            }
            Value::Number(value.sqrt())
        }
        "pow" => Value::Number(
            need_number(interp, name, &args, 0, span)?.powf(need_number(interp, name, &args, 1, span)?),
        ),
        "exp" => Value::Number(one(interp)?.exp()),
        "log" => {
            let value = one(interp)?;
            if args.len() == 1 {
                Value::Number(value.ln())
            } else {
                let base = need_number(interp, name, &args, 1, span)?;
                Value::Number(value.ln() / base.ln())
            }
        }
        "sin" => Value::Number(one(interp)?.sin()),
        "cos" => Value::Number(one(interp)?.cos()),
        "tan" => Value::Number(one(interp)?.tan()),
        "atan2" => Value::Number(
            need_number(interp, name, &args, 0, span)?.atan2(need_number(interp, name, &args, 1, span)?),
        ),
        "hypot" => Value::Number(
            need_number(interp, name, &args, 0, span)?.hypot(need_number(interp, name, &args, 1, span)?),
        ),
        "min" | "max" => {
            let mut best = need_number(interp, name, &args, 0, span)?;
            for index in 1..args.len() {
                let value = need_number(interp, name, &args, index, span)?;
                // NaN wins, as `Math.min`/`Math.max` do: `f64::min` ignores it.
                if value.is_nan() || best.is_nan() {
                    best = f64::NAN;
                } else if (native.method == "min") == (value < best) && value != best {
                    best = value;
                }
            }
            Value::Number(best)
        }
        "clamp" => {
            let value = need_number(interp, name, &args, 0, span)?;
            let low = need_number(interp, name, &args, 1, span)?;
            let high = need_number(interp, name, &args, 2, span)?;
            if low > high {
                return interp.fail(
                    "BAA311",
                    vec![
                        "ram.clamp".into(),
                        "low <= high".into(),
                        "2".into(),
                        "a low bound above the high bound".into(),
                    ],
                    span,
                );
            }
            Value::Number(value.max(low).min(high))
        }
        "lerp" => {
            let a = need_number(interp, name, &args, 0, span)?;
            let b = need_number(interp, name, &args, 1, span)?;
            let t = need_number(interp, name, &args, 2, span)?;
            Value::Number(a + (b - a) * t)
        }
        "idiv" | "modulo" => {
            let a = need_number(interp, name, &args, 0, span)?;
            let b = need_number(interp, name, &args, 1, span)?;
            if b == 0.0 {
                return Err(Flow::Err(
                    interp.error("BAA306", vec![], span).with_note("divisor is zero"),
                ));
            }
            Value::Number(if native.method == "idiv" {
                (a / b).floor()
            } else {
                ((a % b) + b) % b
            })
        }
        "gcd" => {
            let mut a = need_number(interp, name, &args, 0, span)?.abs();
            let mut b = need_number(interp, name, &args, 1, span)?.abs();
            while b > 0.0 {
                let next = a % b;
                a = b;
                b = next;
            }
            Value::Number(a)
        }
        "is_nan" => Value::Bool(match args.first() {
            Some(Value::Number(value)) => value.is_nan(),
            // Anything that is not a number "is not a number".
            _ => true,
        }),
        "is_finite" => Value::Bool(matches!(args.first(), Some(Value::Number(value)) if value.is_finite())),
        "is_whole" => Value::Bool(
            matches!(args.first(), Some(Value::Number(value)) if value.is_finite() && value.fract() == 0.0),
        ),
        "sum" | "mean" | "median" => {
            let items = need_array(interp, name, &args, 0, span)?;
            let mut numbers = Vec::with_capacity(items.len());
            for item in &items {
                match item {
                    Value::Number(value) => numbers.push(*value),
                    other => {
                        return Err(Flow::Err(
                            interp
                                .error(
                                    "BAA311",
                                    vec![name.clone(), "an array of numbers".into(), "1".into(), other.describe()],
                                    span,
                                )
                                .with_note("every item must be a number"),
                        ))
                    }
                }
            }
            match native.method {
                "sum" => Value::Number(numbers.iter().sum()),
                "mean" => {
                    if numbers.is_empty() {
                        Value::Nil
                    } else {
                        Value::Number(numbers.iter().sum::<f64>() / numbers.len() as f64)
                    }
                }
                _ => {
                    if numbers.is_empty() {
                        Value::Nil
                    } else {
                        numbers.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
                        let middle = numbers.len() / 2;
                        Value::Number(if numbers.len() % 2 == 1 {
                            numbers[middle]
                        } else {
                            (numbers[middle - 1] + numbers[middle]) / 2.0
                        })
                    }
                }
            }
        }
        "to_binary" | "to_hex" => {
            let value = one(interp)?;
            if !value.is_finite() || value.fract() != 0.0 {
                return interp.fail(
                    "BAA311",
                    vec![name.clone(), "a whole number".into(), "1".into(), "a fraction".into()],
                    span,
                );
            }
            let magnitude = value.abs() as u64;
            let digits = if native.method == "to_binary" {
                format!("{magnitude:b}")
            } else {
                format!("{magnitude:x}")
            };
            Value::str(if value < 0.0 { format!("-{digits}") } else { digits })
        }
        "parse" => {
            let text = need_string(interp, name, &args, 0, span)?;
            let base = if args.len() > 1 {
                need_number(interp, name, &args, 1, span)?
            } else {
                10.0
            };
            if base == 10.0 {
                match number::parse(&text) {
                    Some(value) => Value::Number(value),
                    None => Value::Nil,
                }
            } else {
                // `parseInt` reads as many valid digits as it can and ignores
                // the rest, which is what this reproduces.
                let trimmed = text.trim();
                let (sign, body) = match trimmed.strip_prefix('-') {
                    Some(rest) => (-1.0, rest),
                    None => (1.0, trimmed.strip_prefix('+').unwrap_or(trimmed)),
                };
                let radix = base as u32;
                if !(2..=36).contains(&radix) {
                    return Ok(Value::Nil);
                }
                let mut total: f64 = 0.0;
                let mut any = false;
                for ch in body.chars() {
                    match ch.to_digit(radix) {
                        Some(digit) => {
                            total = total * base + digit as f64;
                            any = true;
                        }
                        None => break,
                    }
                }
                if any {
                    Value::Number(sign * total)
                } else {
                    Value::Nil
                }
            }
        }
        _ => Value::Nil,
    })
}
